docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,319 @@
+
+#### Kubelet serving certificate rotation daemon
+
+The kubelet serving certificate rotation daemon will run as a systemd or windows service manager service on the node. It will check for the existence of the serving certificate files, and either check their validity and that they are not self signed, or request a new serving certificate. Once the new serving certificate is written, kubelet can be restarted.
+
+#### Core Specification
+- Core Cluster API MUST provide the following implementations of CSRs and signers:
+    - cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure which implement an “Always Allow” type signer that provides equivalent security to Cluster API v1alpha3. This is only to be used for providers where no secure mechanism exists.
+    - cluster.x-k8s.io/kube-apiserver-client-kubelet-tpm and cluster-x-k8s-io/kubelet-serving-tpm
+        - Will implement TPM-based certificate signers and requesters based on the cloud-provider-gcp implementation.
+        - However, since the mechanism for retrieving endorsement keys varies across platforms, the TPM signer will additionally require a provider specific mechanism to provide the TPM Endorsement Key.
+
+#### Provider Specification
+
+#### All providers
+- All providers must insert a ProviderID within the KubernetesNodeProviderIdentifierOID extension attribute of the CSR
+- All signer names must be filled in by the provider’s controller in InfraCluster.Status.KubeletClientCertificateSigner and InfraCluster.Status.KubeletServingCertificateSigner
+
+#### Insecure providers
+- An insecure provider MUST not implement certificate rotation or kubelet serving certificate signing
+- InfraCluster.Status.KubeletClientCertificateSigner MUST be set to cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure
+- An insecure provider MUST use the cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure signer/v1alpha4
+
+#### Secure providers
+- A secure provider MUST implement certificate rotation and kubelet server certificate signing
+- A provider must register signers of:
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-<provider>/v1alpha4`
+    - `cluster-x-k8s-io/kubelet-serving-<provider>/v1alpha4`
+- A secure provider MUST implement a secure attestation mechanism, based upon PEM-encoded blocks within the Certificate Signing Request.
+- Where a secure provider’s attestation mechanism does not include a nonce or timestamp to protect against replay attacks, the mechanism MUST implement a secondary time-limited attestation (e.g. AWS Instance Identity document + AWS HMACv4 signature).
+- A provider’s signer MUST run on the management cluster
+
+#### TPM based providers
+- A TPM provider MUST use the following certificate signers
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-tpm/v1alpha4`
+    - `cluster-x-k8s-io/kubelet-serving-tpm/v1alpha4`
+- A TPM provider MUST annotate new CSRs as follows:
+    - Key: cluster-x-k8s-io/tpm-endorsement-key/v1alpha4
```

Comment:
How do we support multiple versions in the future?

Reply:
I removed this. we don't actually need to version the signers.